cpp/velox/compute/VeloxPlanConverter.cc

```diff
@@ -222,15 +221,48 @@ std::string VeloxPlanConverter::nextPlanNodeId() {
 
 std::shared_ptr<gluten::RecordBatchResultIterator>
 VeloxPlanConverter::GetResultIterator() {
-  std::shared_ptr<gluten::RecordBatchResultIterator> resIter;
+  std::vector<std::shared_ptr<gluten::RecordBatchResultIterator>> inputs = {};
+  return GetResultIterator(inputs);
+}
+
+std::shared_ptr<gluten::RecordBatchResultIterator> VeloxPlanConverter::GetResultIterator(
+    std::vector<std::shared_ptr<gluten::RecordBatchResultIterator>> inputs) {
+  if (inputs.size() > 0) {
+    arrowInputIters_ = std::move(inputs);
+  }
   const std::shared_ptr<const core::PlanNode> planNode = getVeloxPlanNode(plan_);
 
   auto splitInfos = subVeloxPlanConverter_->splitInfos();
-  auto leafPlanNodeIds = planNode->leafPlanNodeIds();
-  // Here only one leaf node is expected here.
-  assert(leafPlanNodeIds.size() == 1);
-  auto iter = leafPlanNodeIds.begin();
-  auto splitInfo = splitInfos[*iter].get();
+  if (splitInfos.size() == 0) {
+    // Source node is not required.
+    auto wholestageIter =
+        std::make_shared<WholeStageResIterMiddleStage>(pool_, planNode, fakeArrowOutput_);
+    return std::make_shared<gluten::RecordBatchResultIterator>(std::move(wholestageIter));
+  }
+
+  // Source node is required.
+  std::vector<std::shared_ptr<facebook::velox::substrait::SplitInfo>> scanInfos;
+  std::vector<core::PlanNodeId> scanIds;
+  for (const auto& leafPlanNodeId : planNode->leafPlanNodeIds()) {
+    if (splitInfos.find(leafPlanNodeId) == splitInfos.end()) {
+      throw std::runtime_error("Could not find leafPlanNodeId.");
+    }
+    auto splitInfo = splitInfos[leafPlanNodeId];
+    if (!splitInfo->isStream) {
+      scanInfos.emplace_back(splitInfo);
+      scanIds.emplace_back(leafPlanNodeId);
+    }
+  }
+  if (scanInfos.size() == 0) {
+    // Source node is not required.
+    auto wholestageIter =
+        std::make_shared<WholeStageResIterMiddleStage>(pool_, planNode, fakeArrowOutput_);
+    return std::make_shared<gluten::RecordBatchResultIterator>(std::move(wholestageIter));
+  }
+  if (scanInfos.size() > 1) {
+    throw std::runtime_error("Only one scan is supported.");
```

Comment:
Can we support multiple scan nodes?

Reply:
Do you refer to the case of build side and stream side both containing scan? I think it can be supported if needed. Velox also supports this pattern. @zzcclp Is there any case like this in your tests?

Velox unit test:
https://github.com/facebookincubator/velox/blob/main/velox/exec/tests/HashJoinTest.cpp#L373-L390

Reply:
Got it, thanks. Will take a look.

Reply:
Fixed this issue on my env, would you like to help verify?